use-cases/client-server-msgs-perf/svmsg_file_server.c

```diff
@@ -340,54 +341,68 @@ main(int argc, char *argv[])
             // misleading. It needs to be hard-coded to a literal string at
             // compile-time. The metric's name given below is the right one
             // corresponding to the active clock.
-            size_t throughput = (uint64_t)
-                        ((clientp->num_ops * 1.0 / clientp->cumu_time_ns)
-                                    * L3_NS_IN_SEC);
+            size_t throughput = 0;
 
             printf("Server: Client ID=%d exited. num_ops=%" PRIu64 " (%s)"
-                   ", cumu_time_ns=%" PRIu64
+                   ", cumu_time_ns=%" PRIu64 " (%s ns)"
                    " (clock_id=%d)"
                    ", Avg. %s time=%" PRIu64 " ns/msg"
                    ", Server-throughput=%lu (%s) ops/sec"
                    ", # active clients=%d\n",
                    req.clientId,
                    clientp->num_ops, value_str(clientp->num_ops),
-                   clientp->cumu_time_ns,
+                   clientp->cumu_time_ns, value_str(clientp->cumu_time_ns),
                    clock_id, time_metric_name(clock_id),
                    (clientp->cumu_time_ns / clientp->num_ops),
                    throughput, value_str(throughput),
                    NumActiveClients);
 
-            if (NumActiveClients == 0) {
-                break;
-            }
             // Client has exited, so no need to send ack-response back to it.
             req.clientId = 0;
+            if (NumActiveClients == 0) {
+                goto end_forever_loop;
+            }
             break;
 
-          case REQ_MT_DECR:
-          case REQ_MT_GET_CTR:
           case REQ_MT_QUIT:
           default:
             assert(1 == 0);
             break;
         }
         // Client may have exited, so no need to send ack-response back to it.
         if (req.clientId && msgsnd(req.clientId, &resp, RESP_MSG_SIZE, 0) == -1) {
+            printf("Warning: msgsnd() to client ID=%d failed to deliver.\n",
+                   req.clientId);
             break;
         }
 
         /* Parent loops to receive next client request */
     }
+
+end_forever_loop:
+    if (clock_gettime(clock_id, &ts1)) {        // **** Timing ends
+        errExit("clock_gettime-ts1");
+    }
     assert(NumActiveClients == 0);
 
+    uint64_t nsec0 = timespec_to_ns(&ts0);
+    uint64_t nsec1 = timespec_to_ns(&ts1);
+    uint64_t elapsed_ns = (nsec1 - nsec0);
+
     if (msgctl(serverId, IPC_RMID, NULL) == -1) {
         errExit("msgctl");
     }
     printf("Server: # active clients=%d (HWM=%d). Exiting.\n",
            NumActiveClients, NumActiveClientsHWM);
 
-    printSummaryStats(ActiveClients, NumActiveClientsHWM, clock_id);
+    printSummaryStats(run_descr, ActiveClients, NumActiveClientsHWM, clock_id,
+                      elapsed_ns);
+
+    // For visibility into how clocks are performing on user's machine,
+    // run clock-calibration after all workload / metrics collection is done.
+#if !defined(L3_ENABLED)
+    svr_clock_calibrate();
```

Comment:
Relocate here to avoid perturbing CPU-clocking and perf-metrics while workload is running.